src/apiv2.ts

```diff
@@ -0,0 +1,283 @@
+  }
+
+  private async getAccessToken(): Promise<string> {
+    // Runtime fetch of Auth singleton to prevent circular module dependencies
+    if (accessToken) {
+      return accessToken;
+    }
+    const data = await require("./auth").getAccessToken(refreshToken, []);
+    return data.access_token;
+  }
+
+  private async doRequest<ReqT, ResT>(
+    options: ClientRequestOptions<ReqT>
+  ): Promise<ClientResponse<ResT>> {
+    if (!options.path.startsWith("/")) {
+      options.path = "/" + options.path;
+    }
+
+    const versionPath = this.opts.apiVersion ? `/${this.opts.apiVersion}` : "";
+    let fetchURL = `${this.opts.urlPrefix}${versionPath}${options.path}`;
+    if (options.queryParams) {
+      // TODO(bkendall): replace this half-hearted implementation with
+      // URLSearchParams when on node >= 10.
+      const sp: string[] = [];
+      for (const key of Object.keys(options.queryParams)) {
+        const value = options.queryParams[key];
+        sp.push(`${key}=${encodeURIComponent(value)}`);
+      }
+      if (sp.length) {
+        fetchURL += "?" + sp.join("&");
+      }
+    }
+
+    const fetchOptions: RequestInit = {
+      headers: options.headers,
+      method: options.method,
+    };
+
+    if (options.json) {
+      fetchOptions.body = JSON.stringify(options.json);
+    }
+
+    this.logRequest(options);
+
+    let res: Response;
+    try {
+      res = await fetch(fetchURL, fetchOptions);
+    } catch (err) {
+      throw new FirebaseError(`Failed to make request to ${fetchURL}`, { original: err });
+    }
+
+    let body: ResT;
+    if (options.responseType === "json") {
+      body = await res.json();
+    } else {
+      // This is how the linter wants the casting to T to work.
+      body = ((await res.text()) as unknown) as ResT;
+    }
+
+    this.logResponse(res, body, options);
+
+    if (res.status >= 400) {
+      throw responseToError({ statusCode: res.status }, body);
+    }
+
+    return {
+      status: res.status,
+      response: res,
+      body,
+    };
+  }
+
+  private logRequest(fetchOptions: ClientRequestOptions<unknown>): void {
+    let searchParamLog = "";
+    if (fetchOptions.queryParams) {
+      if (!fetchOptions.skipLog?.queryParams) {
+        searchParamLog = JSON.stringify(fetchOptions.queryParams);
+      } else {
+        searchParamLog = "[SEARCH PARAMS OMITTED]";
+      }
+    }
+    const versionPath = this.opts.apiVersion ? `/${this.opts.apiVersion}` : "";
+    const urlLog = `${this.opts.urlPrefix}/${versionPath}${fetchOptions.path}`;
+    logger.debug("[apiv2] HTTP REQUEST:", fetchOptions.method, urlLog, searchParamLog);
+    if (fetchOptions.json) {
+      if (!fetchOptions.skipLog?.body) {
+        logger.debug("[apiv2] HTTP REQUEST BODY:", JSON.stringify(fetchOptions.json));
+      } else {
+        logger.debug("[apiv2] HTTP REQUEST BODY OMITTED");
+      }
+    }
+  }
+
+  private logResponse(res: Response, body: unknown, options: ClientRequestOptions<unknown>): void {
```

Comment:
I think we should tag each log line about a request with either the full URL so it's easy to correlate with requests. I also liked the `>>>` and `<<<` personally (helps me visually scan for reqs/responses). How about:

    >>> [v2][headers] POST https://the.url.goes.here/foo/bar {headers: 'go here'}
    >>> [v2][body] POST https://the.url.goes.here/foo/bar {headers: 'go here'}
    <<< [v2][status] POST https://the.url.goes.here/foo/bar 200
    <<< [v2][body] POST https://the.url.goes.here/foo/bar {response: 'body here'}

...or something

Reply:
I can get behind this idea. I did a bit of cleaning up of the log logic so hopefully it's a little easier to come back to in the future.